local_vars/local_vars.cc

```diff
@@ -226,35 +232,246 @@ class LocalNameInserter {
+                                   core::Names::toHashDup(), ast::make_expression<ast::Local>(original.loc, arg.arg));
+                if (!kwArgKeyEntries.empty()) {
+                    // TODO(aprocter): it might make more sense to replace this with an InsSeq that calls
+                    // <Magic>::<merge-hash>, which is what's done in the desugarer.
+                    kwArgsHash = ast::MK::Send1(
+                        original.loc,
+                        ast::MK::Hash(original.loc, std::move(kwArgKeyEntries), std::move(kwArgValueEntries)),
+                        core::Names::merge(), std::move(kwArgsHash));
+                    kwArgKeyEntries.clear();
+                    kwArgValueEntries.clear();
+                }
+            } else if (arg.flags.block) {
+                blockArg = ast::make_expression<ast::Local>(original.loc, arg.arg);
+            } else if (arg.flags.shadow) {
+                ENFORCE(false, "Shadow only come from blocks, but super only looks at a method's args");
             } else {
-                if (auto e = ctx.beginError(original.loc, core::errors::Namer::SelfOutsideClass)) {
-                    e.setHeader("`{}` outside of method", "super");
+                ENFORCE(false, "Unhandled arg kind in ZSuperArgs");
+            }
+        }
+
+        // At this stage the method should always have a block arg, even if it's synthetic (i.e., wasn't mentioned in
+        // the original source).
+        ENFORCE(blockArg, "Block argument not present");
+
+        // If there were any posargs after a positional splat, fold them into the splatted array.
+        if (posArgsArray && !posArgsEntries.empty()) {
+            posArgsArray = ast::MK::Send1(original.loc, std::move(posArgsArray), core::Names::concat(),
+                                          ast::MK::Array(original.loc, std::move(posArgsEntries)));
+            posArgsEntries.clear();
+        }
+
+        // If there were any keyword args after a keyword splat, fold them into the splatted hash.
+        if (kwArgsHash && !kwArgKeyEntries.empty()) {
+            kwArgsHash =
+                ast::MK::Send1(original.loc, std::move(kwArgsHash), core::Names::merge(),
+                               ast::MK::Hash(original.loc, std::move(kwArgKeyEntries), std::move(kwArgValueEntries)));
+            kwArgKeyEntries.clear();
+            kwArgValueEntries.clear();
+        }
+
+        auto method = ast::MK::Literal(
+            original.loc, core::make_type<core::LiteralType>(core::Symbols::Symbol(), core::Names::super()));
+
+        if (posArgsArray) {
+            ast::Send::ARGS_store sendargs;
+            // We wrap self with T.unsafe in order to get around the requirement for <call-with-splat> and
+            // <call-with-splat-and-block> that the shapes of the splatted hashes be known statically. This is a bit of
+            // a hack, but 'super' is currently treated as untyped anyway.
```

Comment:
Wanted to call this out in a comment to make sure it doesn't get buried. I think this is a reasonable thing to do, but open to other suggestions.

Reply:
Can we detect if we'd need to do this? For example, some cases where we don't need to do this:

- The method only has positional args
- (i think) The method doesn't have any `*args` or `**kwargs`

Basically, write some code to detect, based on the for loop above, whether something that is not a shape or tuple will have been constructed.

If we ever get around to fixing #1068 for classes (not modules), it will be nice to not have to remember to change this code.

Reply:
I _think_ that as it stands this will only `unsafe` exactly when we need to in order to placate `call-with-splat{,-and-block}`. (We'll only generate these if `*args` is present.) Will add a comment to that effect.

An alternative (probably best in a separate PR?) would be what's mentioned in this comment (basically, make `call-with-splat` and `call-with-splat-and-block` aware that `super` is special, which we may need to do for non-`zsuper`-synthesized things anyway): https://github.com/sorbet/sorbet/pull/4654#pullrequestreview-766089106 . We could then delurk the unsafe here. That still leaves us with work to find when fixing #1068, but maybe not in such a weird place.